### PR TITLE
AI Completions

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
Added a special escape hatch in the searcher to run AI completions. Screencast was posted previously on discord. You'll need to export `OPENAI_API_KEY` to use the feature, otherwise it's a no-op. Currently only works with tables, more prompts need to be written. Also a few new issues opened to enable nicer prompts in the future.

### Important Notes
The whole searcher-side of things is a mess and it's only a PoC, so don't spend too much time reviewing it, think about how to make it a feature instead :P 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/gui/src/controller/searcher.rs [hunk 1]
```diff
@@ -288,7 +300,7 @@ impl ParsedInput {
                 chain.wrapped.args.push(argument);
             }
             Some(chain)
-        // If there isn't any expression part, the pattern is the whole input.
+            // If there isn't any expression part, the pattern is the whole input.
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: I believe this is an accidental change :)

File: app/gui/src/controller/searcher.rs [hunk 2]
```diff
@@ -720,12 +730,64 @@ impl Searcher {
         self.notifier.notify(Notification::NewActionList);
     }
 
+    const AI_QUERY_PREFIX: &'static str = "AI:";
+    const AI_QUERY_ACCEPT_TOKEN: &'static str = "#";
+    const AI_STOP_SEQUENCE: &'static str = "`";
+    const AI_GOAL_PLACEHOLDER: &'static str = "__$$GOAL$$__";
+
+    async fn accept_ai_query(
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs:578]: I know this is PoC, but docs are needed here for future refactoring/development. Please describe shortly what this function does and why.

File: app/gui/src/controller/searcher.rs [hunk 3]
```diff
@@ -720,12 +730,64 @@ impl Searcher {
...
+        let next = result.next().await.ok_or(NoAIVisualizationDataReceived)?;
+        let prompt = std::str::from_utf8(&next)?;
+        let prompt_with_goal = prompt.replace(Self::AI_GOAL_PLACEHOLDER, &query);
+        graph.detach_visualization(vis.id).await?;
+        let completion = graph.get_ai_completion(&prompt_with_goal, Self::AI_STOP_SEQUENCE).await?;
+        notifier.publish(Notification::AISuggestionUpdated(completion)).await;
+        Ok(())
+    }
+
+    fn handle_ai_query(&self, query: String) -> FallibleResult {
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs:605]: I know this is PoC, but docs are needed here for future refactoring/development. Please describe shortly what this function does and why.

**Review by wdanilo — APPROVED** [review]
QA needed. After QA is done, lets merge it and lets make some hype about it in our promo videos :D

**Review by hubertp — APPROVED** [review]
The engine part looks OK, but I'm guessing you might need to adapt it given recent changes

**Comment by wdanilo** [comment]
@kustosz could you look at the @hubertp comment please and give us a green light when it will be fixed, please? :)

**Comment by xvcgreg** [comment]
@kustosz please take a look at this PR

**Comment by wdanilo** [comment]
Ok @xvcgreg @sylwiabr this is ready for QA according to Marcin :) To test it, you need `OPENAI_API_KEY=<THE_KEY> ./run ide start`. @vitvakatu do you think you'd be able to Q/A it ?

**Comment by vitvakatu** [comment]
QA: passed 🍏 :

**Comment by wdanilo** [comment]
Perfect! It has landed <3

**Pull request merged**